File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = 0;

        int l=0, r=0;
        unordered_map<char,int> st;

        while(r<s.size()){

            if(st.find(s[r])!=st.end()){
                if(st[s[r]]>=l){
                    l = st[s[r]] + 1;
                }
            }
               
               len = max(len,r-l+1);
               st[s[r]] = r;
               r++;

        }
        return len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = 0; int l=0, r=0;
        Map<Character,Integer> mp = new HashMap<>();
        while(r<s.length()){
            char c = s.charAt(r);
            if(mp.containsKey(c)){
                if(mp.get(c)>=l){
                    l = mp.get(c) + 1;
                }
            }

            len = Math.max(len,r-l+1);
            mp.put(c,r);
            r++;
        } 
        return len;   
    }
}